mbus-api/src/receive.rs

```diff
@@ -0,0 +1,166 @@
+                "We have a message from '{}': {:?}",
+                request.sender,
+                request.data
+            );
+            Ok(Self {
+                request,
+                bus_message: bus_msg,
+                reply_type: Default::default(),
+            })
+        } else {
+            Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "invalid message id!",
+            ))
+        }
+    }
+}
+
+/// Message received over the message bus with a reply serialization wrapper
+/// For type safety refer to `ReceivedMessage<'a,S,R>`.
+pub struct ReceivedRawMessage<'a> {
+    bus_msg: &'a nats::asynk::Message,
+}
+
+impl<'a> ReceivedRawMessage<'a> {
+    /// Get a copy of the actual payload data which was sent
+    /// May fail if the raw data cannot be deserialized into `S`
+    pub fn payload<S: Deserialize<'a>>(&self) -> io::Result<S> {
+        let request: SendPayload<S> =
+            serde_json::from_slice(&self.bus_msg.data)?;
+        Ok(request.data)
+    }
+
+    /// Get the identifier of this message.
+    /// May fail if the raw data cannot be deserialized into the preamble.
+    pub fn id(&self) -> io::Result<MessageId> {
+        let preamble: MessagePreamble =
+            serde_json::from_slice(&self.bus_msg.data)?;
+        Ok(preamble.id)
+    }
+
+    /// Respond back to the sender with the `reply` payload wrapped by
+    /// a Result-like type.
+    /// May fail if serialization of the reply fails or if the
+    /// message bus fails to respond.
+    /// Can receive either `R`, `Err()` or `ReplyPayload`.
+    pub async fn respond<T: Serialize, R: Serialize + Into<ReplyPayload<T>>>(
+        &self,
+        reply: R,
+    ) -> io::Result<()> {
+        let reply: ReplyPayload<T> = reply.into();
+        let payload = serde_json::to_vec(&reply)?;
+        self.bus_msg.respond(&payload).await
+    }
+}
+
+impl<'a> std::convert::From<&'a nats::asynk::Message>
+    for ReceivedRawMessage<'a>
+{
+    fn from(value: &'a nats::asynk::Message) -> Self {
+        Self {
+            bus_msg: value,
+        }
+    }
+}
+
+impl<'a, S, R> std::convert::TryFrom<&'a nats::asynk::Message>
+    for ReceivedMessage<'a, S, R>
+where
+    for<'de> S: Deserialize<'de> + 'a + Debug + Clone + MessageIdTrait,
+    R: Serialize,
+{
+    type Error = io::Error;
+
+    fn try_from(value: &'a nats::asynk::Message) -> Result<Self, Self::Error> {
+        ReceivedMessage::<S, R>::new(value)
+    }
+}
+
+impl<'a, S, R> std::convert::TryFrom<ReceivedRawMessage<'a>>
+    for ReceivedMessage<'a, S, R>
+where
+    for<'de> S: Deserialize<'de> + 'a + Debug + Clone + MessageIdTrait,
+    R: Serialize,
+{
+    type Error = io::Error;
+
+    fn try_from(value: ReceivedRawMessage<'a>) -> Result<Self, Self::Error> {
+        ReceivedMessage::<S, R>::new(value.bus_msg)
+    }
+}
+
+impl<T> From<T> for ReplyPayload<T> {
+    fn from(val: T) -> Self {
+        ReplyPayload {
+            0: Ok(val),
+        }
```

Comment:
I wasn't even aware this was legal - but isn't it better to just use ReplyPayload(Ok(val))

Reply:
Yeah, it was originally a "normal" struct, so when I changed the type, the job got half done only :)